### PR TITLE
Launcher converter vm2domain refactor some network parts

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:
Refactor iface multi queue, network indexing and extract SRIOV hostdev creation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by ormergi — CHANGES_REQUESTED** [review]
Hi nice work!

I think we can do better with the errors our code produce, 
sometimes it hard to understand why the code failed and where.

We can improve that to give better explanations and error messages.
Sometimes I find the network related code hard to grasp or lacks context, 
I think that if we would keep the variable names "fully named" and reduce the shortenings and acronyms (at least those that are not common)  it would be easier to understand the code.
for example:
`mq` -> `multiQueue`
`virtioNetMQRequested` -> `IsVirtioNetworkMultuQueueRequested` 
`iface` -> `nic` / `networkInterface` 
...

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 1]
```diff
@@ -1464,30 +1464,9 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
...
-		}
-		if numberOfSources == 0 {
-			return fmt.Errorf("fail network %s must have a network type", network.Name)
-		} else if numberOfSources > 1 {
-			return fmt.Errorf("fail network %s must have only one network type", network.Name)
-		}
-		networks[network.Name] = network.DeepCopy()
+	networks, err := indexNetworksByName(vmi.Spec.Networks)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1468]: The operation description/hint should come from the function/object that created it, otherwise we will have to re-do this in all callers which is a duplication.
The name of the VMI is not a fit this deep in the call stack, it is expected to be outputted to the logger, upper in the stack. I think the virt-launcher is processing everything per a VMI, so this is surely happen already.

The only missing info left now is the "index" operation. Well, I am not sure if this is really needed here. The failure in the indexing is due to the network type configuration, adding the name "index" adds little here, but if you really insist I will add it in the original error message inside `indexNetworksByName`.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 2]
```diff
@@ -1484,38 +1484,11 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 			var pciAddr string
 			pciAddr, sriovPciAddresses, err = popSRIOVPCIAddress(iface.Name, sriovPciAddresses)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to configure SRIOV %s: %v", iface.Name, err)
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1483]: Seems the pattern you are trying to advocate here is for the error at the caller to describe what operation failed by describing the called function. The patter which I think fits better is for the function/object to describe itself (it knows best) and the caller is left to just expressing what failed at its abstraction level and adding relevant info which is not provided by the error itself.

In this case, we add the iface name and express that the SRIOV configuration failed.
The details on why it failed is in the error itself.
This seems to work fine here.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 3]
```diff
@@ -1484,38 +1484,11 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
...
-				addr, err := decoratePciAddressField(iface.PciAddress)
-				if err != nil {
-					return fmt.Errorf("failed to configure SRIOV %s: %v", iface.Name, err)
-				}
-				hostDev.Address = addr
-			}
-
-			if iface.BootOrder != nil {
-				hostDev.BootOrder = &BootOrder{Order: *iface.BootOrder}
+				return fmt.Errorf("failed to configure SRIOV %s: %v", iface.Name, err)
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1487]: same

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 4]
```diff
@@ -1793,6 +1766,32 @@ func indexNetworksByName(networks []v1.Network) (map[string]*v1.Network, error)
 	return netsByName, nil
 }
 
+func createSRIOVHostDevice(hostPCIAddress string, guestPCIAddress string, bootOrder *uint) (HostDevice, error) {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1803]: Done

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 5]
```diff
@@ -1793,6 +1766,32 @@ func indexNetworksByName(networks []v1.Network) (map[string]*v1.Network, error)
...
+	hostAddr, err := decoratePciAddressField(hostPCIAddress)
+	if err != nil {
+		return HostDevice{}, err
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1806]: This is already provided by the error itself. We have nothing to add here.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 6]
```diff
@@ -1766,6 +1762,44 @@ func indexNetworksByName(networks []v1.Network) (map[string]*v1.Network, error)
 	return netsByName, nil
 }
 
+func getSRIOVPCIAddresses(ifaces []v1.Interface) map[string][]string {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1765]: That will hurt the abstraction by exposing implementation details.
At its base, we should split the logic from the data source. I think we can improve this when we will relocate the network and the sriov specific code to a different file/package and structure it in a more testable/modular way.

File: pkg/virt-launcher/virtwrap/api/converter_test.go [hunk 7]
```diff
@@ -1396,21 +1396,38 @@ var _ = Describe("Converter", func() {
...
 			It("should be set on the domain spec for a non-SRIOV nic", func() {
+				vmi.Spec.Domain.Devices.Interfaces[0].PciAddress = pciAddress
 				domain := vmiToDomain(vmi, c)
 				Expect(*domain.Spec.Devices.Interfaces[0].Address).To(Equal(expectedPCIAddress))
 
 			})
 			It("should be set on the domain spec for a SRIOV nic", func() {
-				iface := &vmi.Spec.Domain.Devices.Interfaces[0]
-				iface.SRIOV = &v1.InterfaceSRIOV{}
+				const sriovNetName = "sriov"
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter_test.go:1408]: Yes, that is correct. However, the context covers too many details.
My main problem with it was that the processing is not localized and too distributed.

An alternative could have been to collect the environment variable raw data in the context and pass only that part. If there will be a strong objection to this removal, I will use that as an alternative. But it is already happening in the converter itself, so it is not like the function is not touching the enviourment at the moment.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 8]
```diff
@@ -90,7 +90,6 @@ type ConverterContext struct {
 	HotplugVolumes        map[string]v1.VolumeStatus
 	PermanentVolumes      map[string]v1.VolumeStatus
 	DiskType              map[string]*containerdisk.DiskInfo
-	SRIOVDevices          map[string][]string
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/api/converter.go:93]: I'm curious what's the motivation for removing this map from the context?

With this map, we had a separation where the convertor would simply use the information provided to it to construct a libvirt XML, while hardware-related operations, variables lookup, etc... would happen earlier in the manager part.

**Review by vladikr — COMMENTED** [review]
I'm only concerned that with this change we are moving the discovery and handling of sriov pci addresses to the converter, which in my eyes should stay clean from environment-related operations

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 8 above)
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:93]: The motivation is to keep everything close/localized when possible and not have things happening too far apart, making it harder to structure the code.

If we have a goal here not to cause the converter to depend on the env variables or the pod state, then it is not enforced already (in the converter we read now some fs paths and act on them).

What do you think about adding to the context the env variable raw information (or some basic rendering of it) and allowing the converter processing on it? I just want to avoid having the logic of iterating the vmi interfaces and based on that processing the env variables.
I prefer the information to be collected regardless the VMI spec and then in the converter to translate it to the domain correctly.
In later steps I hope to extract the network part out, and just use it from the converter. Then we can test it independently and re-struct it.

Does it make sense?

**Comment by EdDev** [comment]
> I'm only concerned that with this change we are moving the discovery and handling of sriov pci addresses to the converter, which in my eyes should stay clean from environment-related operations

@vladikr @AlonaKaplan , as the SRIOV mapping is debatable, I removed it from this PR and will handle it in a following one.
That one also relates to the SRIOV migration effort, so it will be good to separate it.

If you have time to look at this now again, I will appreciate it.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/api/converter.go:1488]: Yep this what I meant, but I dont expect the a function to describe the called function or what it do.
I want a function to describe what failed at its abstraction level like you said.
This function is very long and it is hard to follow, failed to configure SRIOV is not enough info especially when you have the exect same error message more then one time (line 1492) it make debugging this very hard giving a bit more info will make it much better.

In popSRIOVPCIAddress function case, we can compromise since it returns only one error that give enough information, but then again if there will be a change for popSRIOVPCIAddress that will make it generate another error it will not reflect in the logs what exactly went wrong

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/api/converter.go:1492]: In this case, the function `createSRIOVHostDevice` does not elaborate much about what did it tried to do that failed.
when there will be an error, it will not reflect in the logs that the this function failed due to hostDevice creation for SRIOV interface.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/api/converter.go]: You right we already get error, Its provided by `utils.ParsePciAddress` that called by `decoratePciAddressField`, which is fine.
We can improve the full error message by adding to this error `failed to decorate PCI address` (or something like that just a suggestion)
that will be reflected in the logs, having this and changing the error at the caller of `createSRIOVHostDevice` will make it much easier to understand from the logs where the code failed.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 9]
```diff
@@ -1559,10 +1554,13 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 
 			// if UseEmulation unset and at least one NIC model is virtio,
 			// /dev/vhost-net must be present as we should have asked for it.
+			mq := vmi.Spec.Domain.Devices.NetworkInterfaceMultiQueue
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/api/converter.go:1509]: Can you please use more meaningful name than mq?

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 10]
```diff
@@ -1800,6 +1779,20 @@ func formatDomainIOThreadPin(vmi *v1.VirtualMachineInstance, domain *Domain, c *
 	return nil
 }
 
+func indexNetworksByName(networks []v1.Network) (map[string]*v1.Network, error) {
+	netsByName := map[string]*v1.Network{}
+	for _, network := range networks {
+		switch {
+		case network.Pod != nil && network.Multus != nil:
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/api/converter.go:1760]: Since you moved the code to a method and called it `indexNetworksByName`, I find it strange to have this validation here. I would introduce another method for this validation or change the method name in some way that will reflect it also has this validation.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 11]
```diff
@@ -1794,6 +1767,34 @@ func indexNetworksByName(networks []v1.Network) (map[string]*v1.Network, error)
...
+	hostDev = HostDevice{
+		Source:  HostDeviceSource{Address: hostAddr},
+		Type:    "pci",
+		Managed: "no",
+	}
+
+	if guestPCIAddress != "" {
+		addr, err := decoratePciAddressField(guestPCIAddress)
+		if err != nil {
+			return hostDev, err
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/api/converter.go:1786]: Why do you return the `hostDev1? I would return `nil, err`.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1488]: I do not understand your point here, sorry. This error is returned from `Convert_v1_VirtualMachine_To_api_Domain`, so at its abstraction level it can say "failed convert from vmi to domain" or just follow the existing format of "failed configure SRIOV" as it was before. This portion of the code indeed handles SRIOV and the info added follows the existing usages here.

The suggested text you gave originally tries to express what the called function does, not what its failure caused.
So I am not clear what exactly is the AI on this comment now.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1492]: It will say for example: `failed to configure SRIOV <iface-name>: failed to parse pci address <address>`.
What information is missing here?

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 9 above)
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1509]: It is used to avoid long lines (or breaking lines).
It has no usage beyond these 2 lines, so I find this redundant.

Do I need to use full names here because it is not clear/readable what it represents?

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1760]: Good point, this indeed does 2 things. I will create another function that does the validation.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 11 above)
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1786]: `HostDevice` is not a pointer here. You suggest to return a pointer instead of the structure?

Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/api/converter.go:1786]: yes

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/api/converter.go:1509]: Up to you, I won't insist. I believe full name will be more readable.

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go]: The full name has no meaning here, is it just a temp variable.
I will embed this into an `if` to make sure no one will try to re-use it beyond its scope here. Hope that will resolve this saga.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 11 above)
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go]: Done

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1762]: Done

**Comment by EdDev** [comment]
change: Answered comments.
(unfortunately I messed something up and the push includes a rebase)

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/api/converter.go:1493]: That it failed during hostDevice creation

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1493]: SRIOV is specifying the logical block in which it failed, which seems enough to me.
Based on what you expect here to have, each function used needs to add its name or what it does to the error chain. So a call stack of 20 will have 20 names chained. This is just a backtrace, and this is not how we log errors for sure.
(but if a projects decides that he wants such a thing, there are tooling to do this)

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/api/converter.go:1489]: My point is `Convert_v1_VirtualMachine_To_api_Domain` return the exact same error in two different places,
in case of an error it will be hard to understand where the code failed

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1489]: How will it return the same exact error? The error returned by `popSRIOVPCIAddress` includes unique information about what actually failed and that is chained to the declaration that the SRIOV failed.

**Review by qinqon — CHANGES_REQUESTED** [review]
I think there is a bug at validation function, we should put a unit test for that also.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 12]
```diff
@@ -1803,6 +1756,52 @@ func formatDomainIOThreadPin(vmi *v1.VirtualMachineInstance, domain *Domain, c *
...
+		hostDev.Address = addr
+	}
+
+	if bootOrder != nil {
+		hostDev.BootOrder = &BootOrder{Order: *bootOrder}
+	}
+
+	return hostDev, nil
+}
+
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1804]: I think I answered that [here](https://github.com/kubevirt/kubevirt/pull/4311/commits/1e55f785e91a94fe2a5de56119019718fe945d2a#r547864151).

Wrapping each returned error for every abstraction layer is not practical. There needs to be a balance between a full trace stack and informative info.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 13]
```diff
@@ -1803,6 +1756,52 @@ func formatDomainIOThreadPin(vmi *v1.VirtualMachineInstance, domain *Domain, c *
 	return nil
 }
 
+func validateNetworksTypes(networks []v1.Network) error {
+	for _, network := range networks {
+		switch {
+		case network.Pod != nil && network.Multus != nil:
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/api/converter.go:1762]: You just commented on the line that checks if both Pod and Multus types are used, it should fail. What do I miss?

[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/api/converter.go:1762]: All good you can forget this comment.

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/api/converter.go:1804]: Understood, we may want to introduce some mechanism to print the stack trace.
